### ISSUE TITLE
SFTP/SSH uploader lacks support for modern algorithms

### ISSUE THREAD

**Issue by AgentOak** [ShareX/ShareX#4907]
ShareX version 13.1

I'm trying to have ShareX upload to my server via SFTP, however in the connection test I get the message
`Failed to negotiate key exchange algorithm.`

My SSH server is configured to only allow these very secure key exchange algorithms:
`KexAlgorithms curve25519-sha256,curve25519-sha256@libssh.org,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512`.

Likewise, ShareX lacks supports for the more secure Encrypt-then-MAC algorithms that I configured `hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com` saying `Server HMAC algorithm not found`.

If I add `diffie-hellman-group-exchange-sha256` to `KexAlgorithms` and `hmac-sha2-512,hmac-sha2-256` to `MACs` on the SSH server, then I get `An established connection was aborted by the server.` in the connection test, however the server log shows
`sshd[xxxxx]: Unable to negotiate with xx.xx.xx.xx port xxxxx: no matching host key type found. Their offer: ssh-rsa,ssh-dss [preauth]`.

Neither key types are acceptable, because `ssh-dss` uses DSA keys which are limited to 1024bit and have been considered insecure for many years. And `ssh-rsa` was just [deprecated by OpenSSH due to SHA1 being too insecure](https://www.zdnet.com/article/openssh-to-deprecate-sha-1-logins-due-to-security-risk/). Secure methods which don't use SHA1 are `ssh-ed25519,rsa-sha2-512,rsa-sha2-256`.

Except for the ETM MACs, [SSH.NET supports the modern algorithms](https://github.com/sshnet/SSH.NET#key-exchange-method), so maybe ShareX is using an outdated SSH.NET version?

**Comment by Jaex** [comment]
> so maybe ShareX is using an outdated SSH.NET version?

We are using latest stable version:

![](https://jaex.getsharex.com/2020/06/devenv_WSSp8zL8MQ.png)

**Comment by AgentOak** [comment]
Oh, I just realized 2020.0 is still in beta. Let's hope they'll be confident to release as stable soon.

**Comment by Telesphoreo** [comment]
So SSH.NET has been updated to 2020.0.0 beta 1 and I still can't upload files via FTP
```Renci.SshNet.Common.SshConnectionException: Server HMAC algorithm not found
   at Renci.SshNet.Session.WaitOnHandle(WaitHandle waitHandle, TimeSpan timeout)
   at Renci.SshNet.Session.Connect()
   at Renci.SshNet.BaseClient.CreateAndConnectSession()
   at Renci.SshNet.BaseClient.Connect()
   at ShareX.UploadersLib.FileUploaders.SFTP.Connect()
   at ShareX.UploadersLib.FileUploaders.SFTP.UploadStream(Stream stream, String remotePath, Boolean autoCreateDirectory)
   at ShareX.UploadersLib.FileUploaders.SFTP.Upload(Stream stream, String fileName)
   at ShareX.WorkerTask.UploadData(IGenericUploaderService service, Stream stream, String fileName)
   at ShareX.WorkerTask.DoUpload(Int32 retry)

Renci.SshNet.Common.SshConnectionException: Server HMAC algorithm not found
   at Renci.SshNet.Session.WaitOnHandle(WaitHandle waitHandle, TimeSpan timeout)
   at Renci.SshNet.Session.Connect()
   at Renci.SshNet.BaseClient.CreateAndConnectSession()
   at Renci.SshNet.BaseClient.Connect()
   at ShareX.UploadersLib.FileUploaders.SFTP.Connect()
   at ShareX.UploadersLib.FileUploaders.SFTP.UploadStream(Stream stream, String remotePath, Boolean autoCreateDirectory)
   at ShareX.UploadersLib.FileUploaders.SFTP.Upload(Stream stream, String fileName)
   at ShareX.WorkerTask.UploadData(IGenericUploaderService service, Stream stream, String fileName)
   at ShareX.WorkerTask.DoUpload(Int32 retry)```
Thought the SSH.NET update would fix it but I guess not? Followed this tutorial and ShareX broke. https://www.ssh-audit.com/hardening_guides.html#ubuntu_20_04_lts

**Comment by AgentOak** [comment]
That guide only allows Encrypt-then-MAC (`etm`) MACs, which are not supported by SSH.NET even in the beta version.

You need to allow the slightly less secure non-etm MACs:
`MACs hmac-sha2-512-etm@openssh.com,hmac-sha2-256-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-512,hmac-sha2-256`

I haven't tried the latest ShareX build myself so not sure if that's enough.

**Comment by Telesphoreo** [comment]
@AgentOak that worked, thanks so much!

**Comment by AgentOak** [comment]
I too can confirm that I can connect to my SSH server with the latest build from AppVeyor.

So fixed by 50c352054a686693612505b178098c5ca99d93d1, apart from the missing support for ETM MACs but that is not a dealbreaker as hmac-sha2-256 is still reasonably secure. There's much more missing in SSH.NET, namely AES-GCM, ChaCha20, umac and rsa-sha2 keys but at least it supports *some* state-of-the-art algorithms now.

**Comment by Jaex** [comment]
Thanks for testing.
